src/Share/Survey/SurveyHelper.cs

```diff
@@ -0,0 +1,360 @@
+            && (Modules[moduleName].ActiveDays == SurveyTriggerCount + 1 && helper.LastPromptDate == Convert.ToDateTime(Modules[moduleName].LastActiveDate) && helper.Today > helper.LastPromptDate.AddDays(DelayForSecondPrompt)
+                || Modules[moduleName].ActiveDays == SurveyTriggerCount + 2 && helper.LastPromptDate == Convert.ToDateTime(Modules[moduleName].LastActiveDate) && helper.Today > LastPromptDate.AddDays(DelayForThirdPrompt));
+
+        private void ModulePromptExpire(SurveyHelper helper, string moduleName, int majorVersion)
+        {
+            helper.Modules[moduleName].ActiveDays = 0;
+        }
+
+        private void MergeScheduleInfo(ScheduleInfo externalScheduleInfo)
+        {
+            DateTime externalLastPromptDate = Convert.ToDateTime(externalScheduleInfo?.LastPromptDate);
+            IDictionary<string, ModuleInfo> externalModules = new Dictionary<string, ModuleInfo>();
+            foreach(ModuleInfo info in externalScheduleInfo?.Modules)
+            {
+                externalModules[info.Name] = info;
+            }
+
+            HashSet<string> moduleNames = new HashSet<string>(Modules.Keys);
+            moduleNames.UnionWith(new HashSet<string>(externalModules.Keys));
+
+            foreach (string name in moduleNames)
+            {
+                if (externalModules.ContainsKey(name) && (!Modules.ContainsKey(name) || Convert.ToDateTime(Modules[name].LastActiveDate) < Convert.ToDateTime(externalModules[name].LastActiveDate)))
+                {
+                    if (externalLastPromptDate != DateTime.MinValue && Convert.ToDateTime(externalModules[name].LastActiveDate) == externalLastPromptDate)
+                    {
+                        LastPromptDate = externalLastPromptDate;
+                    }
+                    Modules[name] = new ModuleInfo(externalModules[name]);
+                }
+            }
+        }
+
+        private ScheduleInfo GetScheduleInfo()
+        { 
+            return new ScheduleInfo() { LastPromptDate = LastPromptDate.ToString("yyyy-MM-dd"), Modules = Modules.Values.ToList() };
+        }
+
+        private bool ReadFromStream()
+        {
+            StreamReader sr = null;
+            try
+            {
+                if (File.Exists(SurveyScheduleInfoFile))
+                {
+                    sr = new StreamReader(new FileStream(SurveyScheduleInfoFile, FileMode.Open, FileAccess.Read, FileShare.None));                    
+                    MergeScheduleInfo(JsonConvert.DeserializeObject<ScheduleInfo>(sr.ReadToEnd()));
+                }
+            }
+            catch (Exception e)
+            {
+                if (e is UnauthorizedAccessException)
+                {
+                    this.IgnoreSchedule = true;
+                }
+                //deserialize failed, means content of file is incorrect, make file empty
+                if (e is JsonException)
+                {
+                    if (sr != null)
```

Comment:
`sr.Dispose` is done in `finally` already. This is not necessary.

Reply:
this is block is a fallback when serialization failed, it will invoke a new write stream, thus the current read stream has to be disposed.